hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/UpgradeDowngrade.java

```diff
@@ -130,6 +133,12 @@ public boolean needsUpgrade(HoodieTableVersion toWriteVersion) {
    * @param instantTime current instant time that should not be touched.
    */
   public void run(HoodieTableVersion toVersion, String instantTime) {
+    // Rollback any pending writes and compact before upgrade/downgrade operations
+    HoodieTable table = upgradeDowngradeHelper.getTable(config, context);
+    rollbackFailedWritesAndCompact(table, context, config, upgradeDowngradeHelper, 
```

Comment:
we can't invoke this for all upgrade/downgrade handlers. it depends on the exact from and to version. 

So, we might have to expose two methods like below to the UpgradeHandler. 

```
boolean needsRollbackPendingCommitAndCompact();
```

once we have this, we might wanna do something like below. bcoz, not all upgrade and downgrade handlers need this to be invoked. 

```
UpgradeDowngrade.run(....) {
      loop through from to to version handlers and check if any handler returns true for needsRollbackPendingCommitAndCompact(). 
      if yes, we need to call rollback pending commits and compact. 
      if not, move on 
      .
      .
 
}
```

Reply:
I implemented the feedback based on the notes above, let me know if rough impl looks ok to you